src/rules/valid-expect-in-promise.ts

```diff
@@ -1,197 +1,354 @@
-  node: PromiseChainCallExpression['callee'],
-  context: RuleContext,
-  testFunctionBody: TSESTree.Statement[],
-) => {
-  promiseCallbacks.some(promiseCallback => {
-    if (promiseCallback && isFunction(promiseCallback)) {
-      if (
-        isExpectCallPresentInFunction(promiseCallback.body) &&
-        node.parent.parent &&
-        !isParentThenOrPromiseReturned(node.parent.parent, testFunctionBody)
-      ) {
-        reportReturnRequired(context, node.parent.parent);
+): TSESTree.BlockStatement['body'] => {
+  let parent: TSESTree.Node['parent'] = node;
 
-        return true;
-      }
+  while (parent) {
+    if (parent.type === AST_NODE_TYPES.BlockStatement) {
+      return parent.body;
     }
 
-    return false;
-  });
+    parent = parent.parent;
+  }
+
+  /* istanbul ignore next */
+  throw new Error(
+    `Could not find BlockStatement - please file a github issue at https://github.com/jest-community/eslint-plugin-jest`,
+  );
 };
 
-const isHavingAsyncCallBackParam = (testFunction: FunctionExpression) =>
-  testFunction.params[0] &&
-  testFunction.params[0].type === AST_NODE_TYPES.Identifier;
+const isDirectlyWithinTestCaseCall = (node: TSESTree.Node): boolean => {
+  let parent: TSESTree.Node['parent'] = node;
+
+  while (parent) {
+    if (isFunction(parent)) {
+      parent = parent.parent;
+
+      return !!(
+        parent?.type === AST_NODE_TYPES.CallExpression && isTestCaseCall(parent)
+      );
+    }
 
-export default createRule<unknown[], MessageIds>({
+    parent = parent.parent;
+  }
+
+  return false;
+};
+
+const isVariableAwaitedOrReturned = (
+  variable: TSESTree.VariableDeclarator,
+): boolean => {
+  const body = findFirstBlockBodyUp(variable);
+
+  // it's pretty much impossible for us to track destructuring assignments,
+  // so we return true to bailout gracefully
+  if (!isIdentifier(variable.id)) {
+    return true;
+  }
+
+  return isValueAwaitedOrReturned(variable.id, body);
+};
+
+export default createRule({
   name: __filename,
   meta: {
     docs: {
       category: 'Best Practices',
-      description: 'Enforce having return statement when testing with promises',
+      description:
+        'Ensure promises that have expectations in their chain are valid',
       recommended: 'error',
     },
     messages: {
-      returnPromise:
-        'Promise should be returned to test its fulfillment or rejection',
+      expectInFloatingPromise:
+        "This promise should either be returned or awaited to ensure the expects in it's chain are called",
     },
     type: 'suggestion',
     schema: [],
   },
   defaultOptions: [],
   create(context) {
+    let inTestCaseWithDoneCallback = false;
+    const chains: boolean[] = [];
+
     return {
-      CallExpression(node) {
-        if (
-          !isPromiseChainCall(node) ||
-          (node.parent && node.parent.type === AST_NODE_TYPES.AwaitExpression)
-        ) {
+      CallExpression(node: TSESTree.CallExpression) {
```

Comment:
why the explicit type?

Reply:
No strong reason - it is symmetrical with the `CallExpression:exit` (which requires it otherwise its an `any`) and ensures WebStorm isn't dumb (for some reason a few of its features sometimes don't resolve types as well if you omit this 🤷)